### PR TITLE
speedtest module!

### PR DESCRIPTION
**Pull request by sivel**
##### SUMMARY
A module that produces speedtest.net results via speedtest-cli? Sure, why not!

##### ISSUE TYPE
 - New Module Pull Request

##### COMPONENT NAME
lib/ansible/modules/net_tools/speedtest.py

##### ANSIBLE VERSION

```
2.7
```

##### ADDITIONAL INFORMATION

```

```

### CODE REVIEW

File: lib/ansible/modules/net_tools/speedtest.py [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+except ImportError:
+    HAS_SPEEDTEST = False
+
+from ansible.module_utils.basic import AnsibleModule
+
+
+def run_speedtest(module):
+    s = speedtest.Speedtest(
+        source_address=module.params['source'],
+        secure=module.params['secure'],
```
Comment (maxamillion) [inline lib/ansible/modules/net_tools/speedtest.py:144]: `secure` isn't defined as a module param

Comment (sivel) [reply lib/ansible/modules/net_tools/speedtest.py:144]: Boo, I renamed right before I pushed to be `validate_certs` and apparently sucked at it.

Comment (sivel) [reply lib/ansible/modules/net_tools/speedtest.py:144]: fixed

**Comment by maxamillion** [comment]
shipit

**Review by maxamillion — APPROVED** [review]

**Comment by bregman-arie** [comment]
Out of curiosity, what would be the use case for using such module?

**Comment by jpmens** [comment]
Out of curiosity: how many people would actually use this module and, in particular, how often in automating things? 

dontshipit

**Comment by maxamillion** [comment]
Out of curiosity, why would you want to attempt to block a module from inclusion in Ansible just because you don't personally have an use case for it?

File: lib/ansible/modules/net_tools/speedtest.py [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+        description:
+          - URL of a Speedtest Mini server
+        required: false
+    pre_allocate:
+        default: true
+        description:
+          - Pre-allocate upload data. This option improves performance of
+            upload testing, but comes at the cost of higher memory use
+        type: bool
+    validate_certs:
```
Comment (KellerFuchs) [inline lib/ansible/modules/net_tools/speedtest.py]: The name of the parameter (`validate_certs`) is at odds with what it actually does.  Please rename to something like `secure` or `use_https`.

File: lib/ansible/modules/net_tools/speedtest.py [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+        default: true
+        description:
+          - Perform download test
+        type: bool
+    exclude:
+        default: []
+        description:
+          - List of servers to exclude from testing
+        required: false
+    mini:
```
Comment (KellerFuchs) [inline lib/ansible/modules/net_tools/speedtest.py]: It's unclear from the name what this is (or even that it's supposed to be a URL).  Please rename.

File: lib/ansible/modules/net_tools/speedtest.py [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+            source=dict(required=False, type='str'),
+            validate_certs=dict(type='bool', default=True),
+            timeout=dict(type='int', default=10),
+            pre_allocate=dict(type='bool', default=True),
+        ),
+        mutually_exclusive=[
+            ['servers', 'mini'],
+        ],
+        required_one_of=[
+            ['download', 'upload'],
```
Comment (KellerFuchs) [inline lib/ansible/modules/net_tools/speedtest.py]: Why is one of `download` and `upload` required, given that they both default to true?

File: lib/ansible/modules/net_tools/speedtest.py [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+        ],
+    )
+
+    if HAS_SPEEDTEST:
+        version_tuple = tuple((int(n) for n in speedtest.__version__.split('.')))
+    else:
+        version_tuple = None
```
Comment (KellerFuchs) [inline lib/ansible/modules/net_tools/speedtest.py]: This `else` branch is unecessary cruft, remove it.

**Review by KellerFuchs — CHANGES_REQUESTED** [review]
@sivel I left you a bunch of actionable comments, but could you also clarify what's the expected use-case?

**Comment by KellerFuchs** [comment]
dontshipit

File: lib/ansible/modules/net_tools/speedtest.py (hunk 5 above)
Comment (KellerFuchs) [reply lib/ansible/modules/net_tools/speedtest.py]: Ping?

**Review by KellerFuchs — CHANGES_REQUESTED** [review]

**Comment by maxamillion** [comment]
rebuild_merge

**Review by abadger — DISMISSED** [review]
Hold one moment... Should this be a facts module?

* It doesn't make any changes to the state of the server.
* It returns something that might be considered "information about the server".

**Comment by KellerFuchs** [comment]
@sivel Did you close the PR because you reopened a new one for a facts module, or did you give up?
If it's the latter, would you mind someone picking up your work?

**Comment by sivel** [comment]
@KellerFuchs There was a discussion that this should be a facts module and merging was dependent on me making it a facts module.  A facts module is one that returns `ansible_facts` instead of relying on a user to register the result.

My personal view point is that making this a facts module was complicated.  Because a facts module, automagically registers it's facts, you have to ensure that multiple runs are namespaced uniquely.

For example, let's say I ran a playbook with the following tasks:

```
- speedtest:
- speedtest:
    server: 1234
- speedtest:
    download: false
- speedtest:
    server: 4567
    upload: false
- speedtest:
    server: 1234
    share: true
```

How would I then namespace each of those runs, so that their results are all retrievable.

A hash of the args wouldn't be programmatically determinable in any useful form.  Keying just off of the server wouldn't be reasonable.

Let's say that I wanted to compare 2 duplicate executions with the same options?

```
- speedtest:
    server: 1234
  register: speedtest_result_1

- pause:
    seconds: 30

- speedtest:
    server: 1234
  register: speedtest_result_2

- name: Assert that 2 speedtest results 30 seconds apart are reasonably consistent
  assert:
    that:
      - speedtest_result_1 - speedtest_result_2 < 1000
```

In the end, I didn't have time to debate the finer intricacies of this problem with @abadger as he was the one with the concern, and dismissed my early arguments with:

> That could be said, with even more arguments about other facts modules, I think.. For instance, any random smpling of cloud facts modules...

Which I believe is more an argument against fact modules than for facts modules.

In any case, let me do this.  Let me re-open this issue.  And perhaps a discussion can happen here.

If there are more concerns, and you would like to help advocate and address any of those concerns, I can work to accept any changes you may have in mind.

**Comment by abadger** [comment]
Yeah, perhaps facts modules should be removed.  However, if we do that people will probably want a replacement....  I don't have ideas for that which would address your namespacing concerns.... (Or perhaps the whole idea of extendable fact gathering is flawed.... Or perhaps we need to limit further what we think needs to be a fact module?  In any case, we've told people before that things which operate similarly to this (make no changes to the system... just retrieving information about the system) need to be facts modules... so I wanted to put the breaks on automerge until we could discuss it.  If we decide we should be allowing modules that only return information about the systems that aren't facts modules then I have no objection... it's when it appears there is a double standard for ansible employees versus community contributors that I holler stop for a moment and lets think this through....

Some notes about this module that are in addition to the above:

* Some of the use cases are better served by using speedtest_cli via the command module.  This module is not idempotent and probably never would be (due to measuring of network speed being so variable) so one of the things we say definitely justifies a separate module isn't available here...

* I don't think all the args would need to be namespaced if this were a facts module.  Upload and download, for instance, should merge with any existing returned value.  A case could be made that a parameter like server also doesn't need to be namespaced... it limits the use case of the module if it's not namespaced but the other use cases could be addressed by either explicit use of register for those cases or use of command.  I'm not sure that I would agree with that case but in devil's advocate mode, I do think it deserves to be tossed into the ring for discussion.

**Comment by KellerFuchs** [comment]
@sivel I think this is a great argument for it not being a facts module, then  :)

**Comment by sivel** [comment]
collections

**Pull request closed without merging**